### ISSUE TITLE
Fetch manifests on sub-resource loads

### ISSUE THREAD

**Issue by dhausknecht** [mikewest/origin-policy#22]
**Open question**
Should we apply a manifest to sub-resources the same way as we do to navigation requests/responses?

**Current state**
An origin manifest is fetched on navigational requests, not for sub-resource loads. For the latter it means that when the page with origin `a.com` loads a script from `b.com` the manifest for `b.com` is not fetched (consequently also not applied even if we have it already in cache).

**Pro(s)**
Let's assume `b.com` from the initial example adds HSTS to the manifest file to no longer send HSTS along with every response. As it is the spirit of the mechanism, a sub-resource load no longer has origin wide effect but the manifest file makes configurations for the origin.

**Con(s)**
Similar to `Upgrade-Insecure-Requests` we send along yet another request header with every single request which itself introduces quite some traffic in terms of extra bytes and we do not win much if at all.

**Comment by annevk** [comment]
I think if we don't it becomes very hard to use the manifest for eventual XSRF+XSS isolation policies as you would be able to load subresources without those being protected.

**Comment by dhausknecht** [comment]
I agree but I didn't want to bias through the way I ask the question ;-) The con came up through a discussion in the office.

**Follow up question**
Would it make sense to ask each origin only once per page load for the manifest?

Then the "we send something every single time" concern would be addressed. But it probably over-complicates things I assume.

**Comment by annevk** [comment]
It seems somewhat reasonable that the origin can say it won't change, and the client caches that and won't ask for a maximum of 24h, somewhat similar to service workers.

@jakearchibald might be able to help out with the details.

**Comment by jakearchibald** [comment]
Capping the max-age of the manifest to 24hrs seems like a good idea, and I'll create a separate issue for that.

Allowing support for Sec-Origin-Policy on subresources seems fine, but I'm not sure we need a client-level cache. Doesn't regular HTTP caching handle this? (as with preflights)

**Comment by annevk** [comment]
The caching would be for not sending the request header for max 24h so we don't bloat requests.

**Comment by jakearchibald** [comment]
Doesn't HPACK prevent the bloating?

**Comment by annevk** [comment]
HPACK doesn't make headers free, no. It just makes them smaller.

**Comment by mikewest** [comment]
I think we must have been talking about different things on Friday. :) I agree completely that we ought to apply the manifest's rules to outgoing requests (subresources, navigations, etc). This is a necessary component of things like the CORS configurations, and is really the entire point of the feature.

The question I thought you were asking was limited to whether we should advertise support for manifests in general on subresource requests (by sending a `Sec-Origin-Policy` header), and load manifests based on subresource responses. I've been thinking about _loading_ manifests as a navigation-only mechanism, but it's quite possible that's overly simplistic.

I don't really understand the discussion of caching. The way I've been thinking about this is that a server's `Sec-Origin-Policy: abc` response leads to a request to `https://example.com/.well-known/origin-policy/abc`, which can deliver response headers that control its maximum lifetime. The policy will continue to be applied to the origin until either the resource expires from the cache (as per the headers just mentioned), or until it's explicitly replaced by a new manifest (`Sec-Origin-Policy: xyz`), or the policy is cleared out (`Sec-Origin-Policy: 0`). All of that seems somewhat orthogonal to whether or not we send the request header.

**Comment by dhausknecht** [comment]
Not sending the header might also confuse servers thinking the feature is not supported yet (let's assume servers handle each request kind of isolated). The result would be sending all the headers in the response. That brings us back to status quo. Not what we want really.
Maybe there is simply no way to avoid sending the `Sec-Origin-Policy` header with each request, including sub-resource requests.

**Comment by domenic** [comment]
My current plan, as reflected in the explainer and version negotiation documents, is to allow any request (including subresource requests) to ask for origin policy updates. So I'll close this, although there is of course still the larger issue of updating the spec to match the explainer.

There's a separate question, discussed in this thread, as to whether we add a header to each request advertising origin policy support. This is now https://github.com/WICG/origin-policy/issues/51.
